src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,261 @@
+    const { time, done, failed } = payload;
+    if (time === this._thumbnailVideoElement.currentTime) {
+      return done(time);
+    }
+    const period = this._manifest.getPeriodForTime(payload.time);
+    if (period === undefined ||
+        period.adaptations.video === undefined ||
+        period.adaptations.video.length === 0) {
+      return failed(new VideoThumbnailLoaderError("NO_TRACK",
+                                                  "Couldn't find track for this time."));
+    }
+    const videoAdaptation = period.adaptations.video[0];
+    const representation = videoAdaptation.trickModeTrack?.representations[0] ??
+                           videoAdaptation.representations[0];
+    if (representation === undefined) {
+      return failed(new VideoThumbnailLoaderError("NO_TRACK",
+                                                  "Couldn't find track for this time."));
+    }
+    const thumbnailTrack = getThumbnailTrack(representation);
+    if (thumbnailTrack.initURL === "") {
+      return failed(new VideoThumbnailLoaderError("NO_INIT_DATA", "No init data for track."));
+    }
+    const thumbnails = getWantedThumbnails(thumbnailTrack,
+                                           payload.time,
+                                           videoElement.buffered);
+    if (thumbnails === null) {
+      return failed(new VideoThumbnailLoaderError("NO_THUMBNAILS", "Couldn't find thumbnail."));
+    }
+    if (thumbnails.length === 0) {
+      log.debug("VTL: Thumbnail already loaded.");
+      return done(time);
+    }
+    log.debug("VTL: Found thumbnails for time", payload.time, thumbnails);
+
+    if (this._currentJob === undefined) {
+      return this.startJob(thumbnailTrack, time, thumbnails, done, failed);
+    }
+
+    if (this._currentJob.thumbnailTrack.codec !== thumbnailTrack.codec ||
+        this._currentJob.thumbnails.length !== thumbnails.length) {
+      this._currentJob.stop();
+      return this.startJob(thumbnailTrack, time, thumbnails, done, failed);
+    }
+    for (let j = 0; j < thumbnails.length; j++) {
+      if (this._currentJob.thumbnails[j].start !== thumbnails[j].start ||
+          this._currentJob.thumbnails[j].duration !== thumbnails[j].duration ||
+          this._currentJob.thumbnails[j].mediaURL !== thumbnails[j].mediaURL) {
+        this._currentJob.stop();
+        return this.startJob(thumbnailTrack, time, thumbnails, done, failed);
+      }
+    }
+    return failed(new VideoThumbnailLoaderError("ALREADY_LOADING",
+                                                "Already loading these thumbnails."));
+  }
+
+  private startJob = (thumbnailTrack: IThumbnailTrack,
+                      time: number,
+                      thumbnails: IThumbnail[],
+                      done: (time: number) => void,
+                      failed: (e: Error) => void
+  ) => {
+    const subscription = getSourceBuffer$(thumbnailTrack,
+                                          this._thumbnailVideoElement).pipe(
```

Comment:
There are two problems here:
  - if the init segment is not the same (e.g. for another period) we will add another mediasource+sourcebuffer thus erasing all thumbnails we already had
  - if the codec is not the same, same things

Both could be done through the exact same SourceBuffer. And all that should already be managed by the QueuedSourceBuffer

Reply:
Done